File: myCut.py
# -*- coding: utf-8 -*-
"""
Created on Thu Jul 26 10:44:54 2018

@author: Wei Yuxuan

Original MATLAB version: myCut.m

Digitization of Flow Measurement Device by CV method
"""

import numpy as np
from PIL import Image
from skimage import filters
#import matplotlib.pyplot as plt

img = Image.open("11.jpg")
f = np.array(img)
h = abs(.5*f[:,:,0] + .5*f[:,:,1] - f[:,:,2])
#img = Image.fromarray(h)
#img.show()
thr = filters.threshold_otsu(h)
hh = (h>=thr)*255                   # 255 for visualization
#img = Image.fromarray(hh)
#img.show()                          # show the binarized
max1 = hh.sum(0).max()
max2 = hh.sum(1).max()
if max1 > max2:
    hh = hh[::-1]
    hh = hh.T
sumH = hh.sum(0)
H1 = sumH[sumH>0]
len_gap = int(len(H1)*65/1755)
left_gap = H1[0:len_gap].sum()
right_gap = H1[(len(H1)-len_gap):len(H1)].sum()
if left_gap < right_gap:
    H1 = H1[::-1]               # 一维数组左右翻转，二维矩阵上下翻转
H1 = H1[0:(len(H1)-len_gap)]
#plt.plot(H1)
const_mark_ratio = 1690
const_mark = np.array([151,196,239,283,326,369,413,456,499,541,583,624,662,701,740,
    779,818,857,896,935,973,1061,1150,1238,1326,1414,1503,1591])
const_mark = const_mark/const_mark_ratio
const_ball_ratio = 54.5161
ball_len = int(len(H1)/const_ball_ratio)
mymin = H1.max()*ball_len
ball_ind = -1
#Lmyball = np.zeros(len(H1)-ball_len)
for i in range(len(H1)-ball_len):
    myball = H1[i:(i+ball_len)].sum()
#    Lmyball[i] = myball
    if myball < mymin:
        ball_ind = i
        mymin = myball
#plt.plot(Lmyball)
ball_ind = ball_ind/len(H1)
cnt = 28
res = -1
for i in const_mark:
    if ball_ind < i:
        res = cnt
        break
    cnt = cnt - 1
print('between mark '+str(res)+' and mark '+str(res+1))

File: myCut.m
clf; close all;
f = imread('8.jpg');		% input image
h = abs(f(:,:,1)-f(:,:,3)); % R通道与B通道绝对值之差
% figure(1);imshow(h);
thr = graythresh(h);
hh = imbinarize(h,thr);     % 自适应二值化
% 自适应调整横竖
max1 = max(sum(hh));
max2 = max(sum(hh,2));
if max1 > max2
    hh = hh';
    hh = fliplr(hh);
end
% figure(2);imshow(hh);
sumH = sum(hh);             % 投影到横轴
% figure(3);plot(sumH);
ind1 = find(sumH>0);
H1 = sumH(sumH>0);
% 自适应调整左右（缺口在右）
len_gap = floor(length(H1)*65/1755);
left_gap = sum(H1(1:len_gap));
right_gap = sum(H1((end-len_gap+1):end));
if left_gap < right_gap
    H1 = fliplr(H1);
end
H1 = H1(1:(end-len_gap+1));  % 去除无关区域
figure(4);plot(H1);         
% 固定刻度的坐标在数组const_mark
diffH = diff(H1);
% figure(5);plot(diffH);
[diff_sort, ind] = sort(diffH);
% mark = [151,499,239,283,779,624,326,456,701,818,1326,369,935,973,1150,...
%     1061,196,1591,541,662,740,857,896,1414,583,1503,413,1238];
% mark = sort(mark);
const_mark_ratio = 1690;
const_mark = [151,196,239,283,326,369,413,456,499,541,583,624,662,701,740,...
    779,818,857,896,935,973,1061,1150,1238,1326,1414,1503,1591];
const_mark = const_mark/const_mark_ratio;   % 归一化
% figure(6);plot(mark,'.');
% 找小球
const_ball_ratio = 54.5161;
ball_len = floor(length(H1)/const_ball_ratio);
mymin = max(H1)*ball_len;
ball_ind = -1;
for i = 1:(length(H1)-ball_len)
    myball = sum(H1(i:(i+ball_len)));
    if myball < mymin
        ball_ind = i;
        mymin = myball;
    end
end
% disp(['ball_ind: ',num2str(ball_ind)]);
ball_ind = ball_ind/length(H1);     % 归一化
% 计算ball_ind在const_mark中的位置
cnt = 28;
res = -1;
for i = const_mark
    if ball_ind < i
        res = cnt;
        break;
    end
    cnt = cnt - 1;
end
disp(['between mark ',num2str(res),' and mark ',num2str(res+1)]);